File: sdks/go/retry.go
// Package client: retry.go implements the opt-in retry policy (#849).
//
// WithRetry applies context-aware, bounded exponential backoff with full
// jitter only where an ambiguous response cannot change the result or
// repeat an unsafe mutation. ErrUnavailable does not prove the server
// rejected the request.
//
// Eligibility is enforced in code, not docs (see requestRetryable):
//
//	reads (Get*/Scan*/Count*/Search*/Illuminate/status)  retryable
//	unconditional PutVertex(es)/PutEdge(s)                retryable
//	plain exact/prefix Deletes, conditional Put, Add      never: original results or
//	                                                     intervening Deletes make replay unsafe
//	receipt-bearing mutations                            never here; the dedicated path verifies
//	                                                     endpoint continuity before each attempt
//	streaming / io (Subscribe/Backup/Restore/…)          never (v1)
//	anything unclassified                                never (fail closed)
//
// Never retried regardless of policy: deterministic outcomes
// (NotFound/InvalidArgument/FailedPrecondition) and DeadlineExceeded (the
// budget is already spent). Default retryable code: Unavailable.
// ResourceExhausted is opt-in via RetryableCodes so it composes with the
// server-side capacity cap (#848) and rate limiter when the caller wants
// retry-after-decay semantics.
package client

import (
	"context"
	"errors"
	"math/rand/v2"
	"time"

	"connectrpc.com/connect"

	pb "github.com/anaregdesign/lantern/pb/graph/v1"
)

// RetryPolicy configures WithRetry. The zero value is normalised to
// MaxAttempts=3, BaseDelay=100ms, MaxDelay=2s, RetryableCodes={Unavailable}.
type RetryPolicy struct {
	// MaxAttempts is the total number of tries including the first.
	// Values < 1 normalise to 3. Under NewLanternFailover the attempts
	// double as the cross-replica budget: an Unavailable attempt advances
	// the sticky endpoint, so the next attempt hits the sibling.
	MaxAttempts int
	// BaseDelay seeds the exponential backoff (default 100ms). The delay
	// before attempt k (k >= 2) is uniformly random in
	// (0, min(MaxDelay, BaseDelay·2^(k-2))] — "full jitter".
	BaseDelay time.Duration
	// MaxDelay caps a single backoff sleep (default 2s).
	MaxDelay time.Duration
	// RetryableCodes lists the connect codes worth another attempt.
	// Empty defaults to {CodeUnavailable}. Add CodeResourceExhausted to
	// retry through the #848 capacity cap / rate limiter.
	RetryableCodes []connect.Code

	// Test seams: injected sleep and randomness. Nil uses real time and
	// math/rand. Unexported — production callers cannot (and must not)
	// touch them.
	sleepFn func(context.Context, time.Duration) error
	randFn  func() float64
}

// normalized returns a copy with defaults applied.
func (p RetryPolicy) normalized() RetryPolicy {
	if p.MaxAttempts < 1 {
		p.MaxAttempts = 3
	}
	if p.BaseDelay <= 0 {
		p.BaseDelay = 100 * time.Millisecond
	}
	if p.MaxDelay <= 0 {
		p.MaxDelay = 2 * time.Second
	}
	if len(p.RetryableCodes) == 0 {
		p.RetryableCodes = []connect.Code{connect.CodeUnavailable}
	}
	if p.sleepFn == nil {
		p.sleepFn = ctxSleep
	}
	if p.randFn == nil {
		p.randFn = rand.Float64
	}
	return p
}

// retryableErr reports whether err is worth another attempt under the
// policy. nil is never retryable (success); DeadlineExceeded/Canceled are
// excluded structurally because the caller's ctx is already dead.
func (p RetryPolicy) retryableErr(err error) bool {
	if err == nil {
		return false
	}
	code := connect.CodeOf(err)
	if code == connect.CodeDeadlineExceeded || code == connect.CodeCanceled {
		return false
	}
	for _, c := range p.RetryableCodes {
		if code == c {
			return true
		}
	}
	return false
}

// delay computes the full-jitter backoff before attempt number `attempt`
// (0-based count of completed attempts): rand(0, min(MaxDelay, Base·2^n)].
func (p RetryPolicy) delay(attempt int) time.Duration {
	ceil := p.BaseDelay << attempt
	if ceil <= 0 || ceil > p.MaxDelay { // <<-overflow guards land on MaxDelay
		ceil = p.MaxDelay
	}
	return time.Duration(p.randFn() * float64(ceil))
}

// run drives attempt() under the policy: first try immediately, then up to
// MaxAttempts-1 retries with jittered backoff. A non-retryable error (or
// success) returns immediately. Backoff sleeps honour ctx; cancellation
// mid-backoff aborts with the last attempt's error joined with ctx.Err().
func (p RetryPolicy) run(ctx context.Context, attempt func() error) error {
	var lastErr error
	for i := 0; i < p.MaxAttempts; i++ {
		if i > 0 {
			if err := p.sleepFn(ctx, p.delay(i-1)); err != nil {
				return errors.Join(lastErr, err)
			}
		}
		lastErr = attempt()
		if !p.retryableErr(lastErr) {
			return lastErr
		}
	}
	return lastErr
}

// ctxSleep sleeps for d or until ctx is done, whichever comes first.
func ctxSleep(ctx context.Context, d time.Duration) error {
	if d <= 0 {
		return ctx.Err()
	}
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-t.C:
		return nil
	}
}

// requestRetryable is the code-enforced eligibility matrix at the wire
// boundary (#849): it classifies by the REQUEST message that is about to
// be (re)sent. ContribIDs only deduplicate while the contribution is live:
// an intervening Delete or expiration makes plain Add unsafe to replay.
// Unknown request types are NOT retryable — fail closed; the paired matrix
// test forces a classification for every public RPC method.
func requestRetryable(req any) bool {
	switch r := req.(type) {
	case *pb.PutVertexRequest:
		return r.GetReceiptContext() == nil && !r.GetIfAbsent()
	case *pb.PutVerticesRequest:
		return r.GetReceiptContext() == nil && !r.GetIfAbsent()
	case *pb.AddEdgeRequest, *pb.AddEdgesRequest,
		*pb.DeleteVertexRequest, *pb.DeleteVerticesRequest,
		*pb.DeleteEdgeRequest, *pb.DeleteEdgesRequest,
		*pb.DeleteVerticesByPrefixRequest, *pb.DeleteEdgesByPrefixRequest:
		return false
	case *pb.GetVertexRequest, *pb.GetVerticesRequest,
		*pb.GetEdgeRequest, *pb.GetEdgesRequest,
		*pb.PutEdgeRequest, *pb.PutEdgesRequest,
		*pb.ScanVerticesRequest, *pb.ScanVertexKeysRequest,
		*pb.ScanEdgesRequest, *pb.CountVerticesByPrefixRequest,
		*pb.SearchVerticesRequest, *pb.IlluminateRequest,
		*pb.GetServerStatusRequest, *pb.GetReplicationStatusRequest,
		*pb.GetReceiptCapabilityRequest, *pb.GetReceiptStatusRequest,
		*pb.GetReceiptStatusesRequest:
		return true
	}
	return false
}

// methodRetryClass is the SDK-method-level eligibility matrix. It exists
// for two consumers: the Failover wrappers (which dispatch per method, not
// per wire request) and the paired reflection test that walks EVERY public
// *Lantern RPC method and fails the build-time contract when a new method
// lands without a classification — the forced-decision rule.
type methodRetryClass int

const (
	// retryNever: never replayed (unsafe writes, streaming, or unclassified).
	retryNever methodRetryClass = iota
	// retryAlways: eligible for replay, including endpoint-bound receipt
	// mutations through their dedicated continuity-aware path.
	retryAlways
)

// methodRetryClasses classifies every public RPC-shaped method on *Lantern
// (and therefore every Failover wrapper). Adding an RPC method without a
// row here fails TestRetryEligibilityMatrix_CoversEveryRPC.
var methodRetryClasses = map[string]methodRetryClass{
	"GetVertex":                      retryAlways,
	"GetVertices":                    retryAlways,
	"GetEdge":                        retryAlways,
	"GetEdges":                       retryAlways,
	"PutVertex":                      retryAlways,
	"PutVertexAt":                    retryAlways,
	"PutVertices":                    retryAlways,
	"PutVertexIfAbsent":              retryNever,
	"PutVertexIfAbsentAt":            retryNever,
	"PutVerticesIfAbsent":            retryNever,
	"PutEdge":                        retryAlways,
	"PutEdgeAt":                      retryAlways,
	"PutEdges":                       retryAlways,
	"DeleteVertex":                   retryNever,
	"DeleteVertices":                 retryNever,
	"DeleteEdge":                     retryNever,
	"DeleteEdges":                    retryNever,
	"DeleteVerticesByPrefix":         retryNever,
	"DeleteEdgesByPrefix":            retryNever,
	"ScanVertices":                   retryAlways,
	"ScanVerticesAll":                retryAlways,
	"ScanVertexKeys":                 retryAlways,
	"ScanVertexKeysAll":              retryAlways,
	"ScanEdges":                      retryAlways,
	"ScanEdgesAll":                   retryAlways,
	"CountVerticesByPrefix":          retryAlways,
	"SearchVertices":                 retryAlways,
	"SearchVerticesPage":             retryAlways,
	"Illuminate":                     retryAlways,
	"GetServerStatus":                retryAlways,
	"GetReplicationStatus":           retryAlways,
	"GetReceiptCapability":           retryAlways,
	"GetReceiptStatus":               retryAlways,
	"GetReceiptStatuses":             retryAlways,
	"PutVertexWithReceipt":           retryAlways,
	"PutVertexAtWithReceipt":         retryAlways,
	"PutVerticesWithReceipt":         retryAlways,
	"PutVertexIfAbsentWithReceipt":   retryAlways,
	"PutVertexIfAbsentAtWithReceipt": retryAlways,
	"PutVerticesIfAbsentWithReceipt": retryAlways,
	"DeleteVertexWithReceipt":        retryAlways,
	"DeleteVerticesWithReceipt":      retryAlways,
	"DeleteEdgeWithReceipt":          retryAlways,
	"DeleteEdgesWithReceipt":         retryAlways,
	"AddEdgeWithReceipt":             retryAlways,
	"AddEdgeAtWithReceipt":           retryAlways,
	"AddEdgesWithReceipt":            retryAlways,
	"Ping":                           retryAlways,
	"AddEdge":                        retryNever,
	"AddEdgeAt":                      retryNever,
	"AddEdges":                       retryNever,
	"AddDecayingEdge":                retryNever, // fans out into an AddEdges batch

	"Backup":               retryNever, // whole-graph stream dump — excluded in v1
	"Restore":              retryNever, // stream restore — excluded in v1
	"Subscribe":            retryNever, // server-streaming replication feed
	"BootstrapIdentity":    retryNever, // recovery stream must not retry invisibly
	"SubscribeIdentity":    retryNever, // caller owns its durable vector cursor
	"NewIncrementalSearch": retryNever, // session constructor; per-query retries ride unary
	"SearchVerticesIter":   retryNever, // local iterator; each page owns its unary retry
}

// retryableMethod is the method-name counterpart to requestRetryable,
// consumed by Failover. Unknown methods fail closed (retryNever is the
// zero value). Receipt-bearing methods use this only on a pinned endpoint.
func retryableMethod(method string) bool {
	return methodRetryClasses[method] == retryAlways
}

File: sdks/dart/lib/src/retry.dart
part of 'client.dart';

/// Opt-in bounded retry policy for mobile network failures.
///
/// The zero/default fields normalize to three total attempts, 100 ms base
/// delay, a 2 second per-delay cap, and retrying `unavailable` only.
final class RetryPolicy {
  /// Creates a retry policy. Supplying this object opts the client into retry.
  const RetryPolicy({
    this.maxAttempts = 0,
    this.baseDelay = Duration.zero,
    this.maxDelay = Duration.zero,
    this.retryResourceExhausted = false,
  });

  /// Total attempts including the first; values below one normalize to three.
  final int maxAttempts;

  /// Exponential backoff seed; non-positive values normalize to 100 ms.
  final Duration baseDelay;

  /// Maximum one backoff delay; non-positive values normalize to 2 seconds.
  final Duration maxDelay;

  /// Whether `resource_exhausted` joins the default `unavailable` status.
  final bool retryResourceExhausted;

  _NormalizedRetryPolicy _normalized() => _NormalizedRetryPolicy(
    maxAttempts: maxAttempts < 1 ? 3 : maxAttempts,
    baseDelay: baseDelay <= Duration.zero
        ? const Duration(milliseconds: 100)
        : baseDelay,
    maxDelay: maxDelay <= Duration.zero ? const Duration(seconds: 2) : maxDelay,
    retryResourceExhausted: retryResourceExhausted,
  );
}

/// A retry-eligible operation exhausted its bounded attempt budget.
final class LanternRetryExhaustedException implements Exception {
  /// Creates an exhausted-attempt wrapper around the final typed cause.
  const LanternRetryExhaustedException({
    required this.attempts,
    required this.cause,
  });

  /// Total attempts performed.
  final int attempts;

  /// Final typed transport failure.
  final LanternException cause;

  @override
  String toString() =>
      'LanternRetryExhaustedException(attempts: $attempts, cause: $cause)';
}

/// Internal retry classes, exposed only through the package's `src` library
/// so the coverage gate can force a decision for every generated RPC.
enum RpcRetryClass {
  /// Unary read/query/status operation.
  read,

  /// Idempotent state-setting write.
  stablePut,

  /// Additive write requiring stable contribution IDs.
  additive,

  /// Observable-result mutation that must not be replayed.
  never,

  /// Streaming operation that must not be replayed.
  stream,
}

/// Fail-closed retry classification registry used by SDK facades and tests.
final class RetryRegistry {
  RetryRegistry._();

  /// Classification for every generated Lantern service method.
  static const Map<String, RpcRetryClass> classifications = {
    'Illuminate': RpcRetryClass.read,
    'GetVertex': RpcRetryClass.read,
    'GetVertices': RpcRetryClass.read,
    'PutVertex': RpcRetryClass.stablePut,
    'PutVertices': RpcRetryClass.stablePut,
    'DeleteVertex': RpcRetryClass.never,
    'DeleteVertices': RpcRetryClass.never,
    'ScanVertices': RpcRetryClass.read,
    'ScanVertexKeys': RpcRetryClass.read,
    'SearchVertices': RpcRetryClass.read,
    'CountVerticesByPrefix': RpcRetryClass.read,
    'DeleteVerticesByPrefix': RpcRetryClass.never,
    'TopVerticesByDegree': RpcRetryClass.read,
    'GetEdge': RpcRetryClass.read,
    'GetEdges': RpcRetryClass.read,
    'AddEdge': RpcRetryClass.additive,
    'AddEdges': RpcRetryClass.additive,
    'PutEdge': RpcRetryClass.stablePut,
    'PutEdges': RpcRetryClass.stablePut,
    'DeleteEdge': RpcRetryClass.never,
    'DeleteEdges': RpcRetryClass.never,
    'DeleteEdgesByPrefix': RpcRetryClass.never,
    'ScanEdges': RpcRetryClass.read,
    'GetServerStatus': RpcRetryClass.read,
    'GetReplicationStatus': RpcRetryClass.read,
    'BackupSnapshot': RpcRetryClass.stream,
    // High-level facades whose result semantics differ from their shared wire
    // request or whose operation expands into AddEdges.
    'PutVertexIfAbsent': RpcRetryClass.never,
    'PutVerticesIfAbsent': RpcRetryClass.never,
    'AddDecayingEdge': RpcRetryClass.additive,
    'ScanVerticesAll': RpcRetryClass.read,
    'ScanVertexKeysAll': RpcRetryClass.read,
    'ScanEdgesAll': RpcRetryClass.read,
  };

  /// Returns [RpcRetryClass.never] for unknown methods.
  static RpcRetryClass classify(String method) =>
      classifications[method] ?? RpcRetryClass.never;

  static bool _allows(String method, {required bool additiveSafe}) {
    return switch (classify(method)) {
      RpcRetryClass.read || RpcRetryClass.stablePut => true,
      RpcRetryClass.additive => additiveSafe,
      RpcRetryClass.never || RpcRetryClass.stream => false,
    };
  }
}

/// Builds Lantern's canonical 24-byte contribution ID.
///
/// The first 16 bytes are [nonce]. The final eight bytes are the big-endian
/// uint64 `(sequence << 16) | index`. Persist the returned bytes with a durable
/// outbox intent when replay must survive process death.
Uint8List contributionIdFrom({
  required Uint8List nonce,
  required BigInt sequence,
  required int index,
}) {
  if (nonce.length != 16) {
    throw _invalidArgumentException('contribution nonce must be 16 bytes');
  }
  final maxSequence = (BigInt.one << 48) - BigInt.one;
  if (sequence < BigInt.zero || sequence > maxSequence) {
    throw _invalidArgumentException('contribution sequence must fit uint48');
  }
  if (index < 0 || index > 0xffff) {
    throw _invalidArgumentException('contribution index must fit uint16');
  }
  final result = Uint8List(24)..setRange(0, 16, nonce);
  var packed = (sequence << 16) | BigInt.from(index);
  for (var offset = 23; offset >= 16; offset--) {
    result[offset] = (packed & BigInt.from(0xff)).toInt();
    packed >>= 8;
  }
  return result;
}

final class _ContribIdGenerator {
  _ContribIdGenerator(this._nonce);

  factory _ContribIdGenerator.secure() {
    final random = Random.secure();
    return _ContribIdGenerator(
      Uint8List.fromList(List.generate(16, (_) => random.nextInt(256))),
    );
  }

  final Uint8List _nonce;
  BigInt _sequence = BigInt.zero;

  List<List<int>?> fillMissing(List<List<int>?> supplied) {
    if (!supplied.any((value) => value == null)) return supplied;
    _sequence += BigInt.one;
    return List<List<int>?>.generate(supplied.length, (index) {
      final existing = supplied[index];
      return existing ??
          contributionIdFrom(nonce: _nonce, sequence: _sequence, index: index);
    }, growable: false);
  }
}

final class _NormalizedRetryPolicy {
  _NormalizedRetryPolicy({
    required this.maxAttempts,
    required this.baseDelay,
    required this.maxDelay,
    required this.retryResourceExhausted,
  });

  final int maxAttempts;
  final Duration baseDelay;
  final Duration maxDelay;
  final bool retryResourceExhausted;
  final Random _random = Random();

  bool retryable(LanternException error) {
    if (error.code == LanternCode.canceled ||
        error.code == LanternCode.deadlineExceeded) {
      return false;
    }
    return error.code == LanternCode.unavailable ||
        (retryResourceExhausted && error.code == LanternCode.resourceExhausted);
  }

  Duration delay(int completedAttempts) {
    var ceilingMicros = baseDelay.inMicroseconds;
    for (var i = 1; i < completedAttempts; i++) {
      if (ceilingMicros >= maxDelay.inMicroseconds ~/ 2) {
        ceilingMicros = maxDelay.inMicroseconds;
        break;
      }
      ceilingMicros *= 2;
    }
    if (ceilingMicros > maxDelay.inMicroseconds) {
      ceilingMicros = maxDelay.inMicroseconds;
    }
    final jittered = (_random.nextDouble() * ceilingMicros).floor();
    return Duration(microseconds: jittered);
  }
}

extension _RetryClient on LanternClient {
  LanternCallOptions? _freezeCallOptions(LanternCallOptions? options) {
    final deadline = options?.deadline;
    if (deadline != null) {
      return LanternCallOptions(
        deadline: deadline,
        cancellation: options?.cancellation,
      );
    }
    final timeout = options?.timeout ?? _defaultTimeout;
    if (timeout == null) {
      return options?.cancellation == null
          ? null
          : LanternCallOptions(cancellation: options?.cancellation);
    }
    return LanternCallOptions(
      deadline: DateTime.now().add(timeout),
      cancellation: options?.cancellation,
    );
  }

  Future<T> _runWithRetry<T>({
    required String method,
    required bool additiveSafe,
    required LanternCallOptions? options,
    required Future<T> Function() attempt,
  }) async {
    final policy = _retryPolicy;
    if (policy == null ||
        !RetryRegistry._allows(method, additiveSafe: additiveSafe)) {
      return attempt();
    }

    LanternException? lastError;
    for (
      var attemptIndex = 0;
      attemptIndex < policy.maxAttempts;
      attemptIndex++
    ) {
      if (attemptIndex > 0) {
        await _waitForRetry(policy.delay(attemptIndex), options);
      }
      try {
        return await attempt();
      } on LanternException catch (error) {
        if (!policy.retryable(error)) rethrow;
        lastError = error;
      }
    }
    throw LanternRetryExhaustedException(
      attempts: policy.maxAttempts,
      cause: lastError!,
    );
  }
}

Future<void> _waitForRetry(Duration delay, LanternCallOptions? options) async {
  final context = _InvocationContext(options, null);
  try {
    await _raceAbort(Future<void>.delayed(delay), context.signal);
  } on connect.ConnectException catch (error) {
    throw _mapConnectException(error, context);
  } finally {
    context.dispose();
  }
}
